modules/guides/src/Controller/GuidesController.php

```diff
@@ -48,69 +63,51 @@ public static function create(ContainerInterface $container) {
    * @return array
    *   Render array.
    */
-  public function listGuides() {
-    $guides = [];
-    $guides_dir = DRUPAL_ROOT . (Settings::get('guides_dir') ?? '/guides');
-
-    foreach (array_diff(scandir($guides_dir), ['..', '.']) as $guide_dir) {
-      $dir = $guides_dir . '/' . $guide_dir;
-      if (is_dir($dir)) {
-        foreach (glob($dir . '/*.md') as $md) {
-          if (file_exists($md)) {
-            $parts = pathinfo($md);
-            $link = Url::fromRoute('guides.guide', ['filename' => $parts['filename']]);
-            $link = Link::fromTextAndUrl(str_replace('_', ' ', $guide_dir), $link)->toString();
-            $guides[] = [
-              '#markup' => $link,
-            ];
-          }
-        }
-      }
-    }
+  public function listGuides(): array {
+    // Get the links of the files.
+    $guides = $this->guidesStorage->getLinks();
 
+    // If the guides array isn't empty, return its content as item list.
     if (!empty($guides)) {
       return [
         '#theme' => 'item_list',
         '#items' => $guides,
       ];
     }
+    // Else return markup text.
     else {
       return [
-        '#markup' => $this->t('<strong>@text</strong>', ['@text' => 'No guides found.']),
+        '#type' => 'html_tag',
+        '#tag' => 'strong',
+        '#value' => $this->t('No guides found.'),
       ];
     }
   }
 
   /**
    * Page callback that renders a markdown file on the UI.
    *
+   * @param string $subdirectory
+   *   The name of the subdirectory which contains the guide file.
+   * @param string $filename
+   *   The name of the guide file.
+   *
    * @return array
    *   Render array.
+   *
+   * @throws \Drupal\guides\Exception\FileNotFoundException
+   *   Thrown when no guide file is found.
    */
-  public function guideContent($filename) {
-    $guides_dir = Settings::get('guides_dir') ?? '/guides';
-    $target = [
-      'dir' => FALSE,
-      'file' => FALSE,
-    ];
-
-    foreach (array_diff(scandir(DRUPAL_ROOT . $guides_dir), ['..', '.']) as $guide_dir) {
-      $dir = DRUPAL_ROOT . $guides_dir . '/' . $guide_dir;
-      $file = $dir . '/' . $filename . '.md';
-      if (file_exists($file)) {
-        $target['dir'] = $guides_dir . '/' . $guide_dir;
-        $target['file'] = $file;
-        break;
-      }
+  public function guideContent(string $subdirectory, string $filename): array {
+    try {
+      // Parses the guide file.
+      $md = new \Parsedown();
+      $md = $md->text($this->guidesStorage->getFileContent($subdirectory, $filename));
     }
-
-    if (!$target['file']) {
-      throw new NotFoundHttpException();
+    catch (FileNotFoundException $e) {
```

Comment:
Why do you catch is and constructs a new exception with the same type again to re-throw it? :open_mouth: 
With my previous comment, I have just wanted to warn you that a `FileNotFoundException` can be thrown and it should be documented in the method's doc block.